contracts/Marketplace.sol

```diff
@@ -1,74 +1,296 @@
+    return payment.purchaser;
+  }
 
-    function requestAccess(bytes32 version) public whenNotPaused {
-        require(!_versionAccess[version][msg.sender], "You already have access to this version of the service");
-        _versionAccess[version][msg.sender] = true;
-        return;
+  // Index
+
+  function getServiceIndex(bytes memory sid) public view returns (uint serviceIndex) {
+    for (uint i = 0; i < services.length; i++) {
+      if (compareBytes(services[i].sid, sid)) {
+        return i;
+      }
     }
+    require(false, "Service not found");
+  }
 
-    function hasAccessToVersion(bytes32 version, address user) public view returns (bool) {
-        return _versionAccess[version][user];
+  function getServiceVersionIndex(uint serviceIndex, bytes20 hash) public view returns (uint versionIndex) {
+    Service storage service = services[serviceIndex];
+    for (uint i = 0; i < service.versions.length; i++) {
+      if (service.versions[i].hash == hash) {
+        return i;
+      }
     }
-}
+    require(false, "Version not found");
+  }
 
-contract Marketplace is Ownable, Pausable {
-    using Address for address;
+  function getServicePaymentIndex(uint serviceIndex, address purchaser) public view returns (uint paymentIndex) {
+    Service storage service = services[serviceIndex];
+    for (uint i = 0; i < service.payments.length; i++) {
+      if (service.payments[i].purchaser == purchaser) {
+        return i;
+      }
+    }
+    require(false, "Payment not found");
+  }
 
-    mapping (bytes32 => address) public serviceContracts;
-    bytes32[] public services;
-    uint256 public totalServices;
-    
-    event ServiceCreated(bytes32 sid, address serviceAddress);
+  // Count
 
-    constructor() Ownable() Pausable() public {}
+  function getServicesCount() public view returns (uint servicesCount) {
+    return services.length;
+  }
 
-    function createService(bytes32 sid) public whenNotPaused returns (address) {
-        if (serviceContracts[sid].isContract()) {
-            return;
-        }
-        serviceContracts[sid] = new Service(msg.sender, sid);
-        services.push(sid);
-        totalServices = totalServices + 1;
-        emit ServiceCreated(sid, serviceContracts[sid]);
-        return serviceContracts[sid];
+  function getServiceVersionsCount(uint serviceIndex) public view returns (uint serviceVersionsCount) {
+    return services[serviceIndex].versions.length;
+  }
+
+  function getServicePaymentsCount(uint serviceIndex) public view returns (uint servicePaymentsCount) {
+    return services[serviceIndex].payments.length;
+  }
+
+  // ------------------------------------------------------
+  // Setter functions
+  // ------------------------------------------------------
+
+  // Manage Service
+
+  function createService (bytes memory sid, uint256 price) public whenNotPaused returns (uint serviceIndex) {
+    require(!isServiceSidExist(sid), "Service's sid is already used");
```

Comment:
we need to find other way to check this, it wil cost a lot of gas in the future and it is time consuming as well. Creating service will become more and more expensive and it should be constant.

services map(sid => bool) - this is perfect usecase for map

Reply:
Yes you are right. Next step is to optimize the searches.